File: Problem2.py
'''

Time Complexity : O(n)
Space Complexity : O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach

Using a hashmap to store the index of the element which have occurred previously. As I loop thru the list, a variable is maintained to track
the maximum length formed without repitions

'''

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char = defaultdict(int)

        i = j = 0
        length = 0

        while j < len(s):
            if s[j] in char:
                i = max(char[s[j]]+1,i)
            char[s[j]] = j

            length = max((j-i+1),length)
            j += 1

        return length

File: Problem2.cs
public class LongestNonRepeatingSubstring
	{
        // Time Complexity : O(n), where n is the length of the string
        // Space Complexity : O(1) - constant space as there could be maximum 26 characters in hashmap
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        public int LengthOfLongestSubstring(string s)
        {
            Dictionary<char, int> map = new Dictionary<char, int>();
            int slow = 0;
            int fast = 0;
            int maxLength = 0;
            for(int i = 0; i < s.Length; i++)
            {
                char c = s[i];
                if (map.ContainsKey(c))
                {
                    slow = Math.Max(slow, map[c] + 1);
                    map[c] = i;
                }
                else
                {
                    map.Add(c, i);
                }
                maxLength = Math.Max(maxLength, i - slow + 1);
            }
            return maxLength;
        }
    }
